include/aws/crt/ByteBuf.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/crt/Outcome.h>
+#include <aws/crt/Types.h>
+
+#include <aws/common/byte_buf.h>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        class ByteCursor
```

Comment:
There's definitely some tension between idiomatic and thin-wrapper here.  Ultimately, we're not just servicing C++ end-users, we're also servicing those people (mostly ourselves, to be honest) who are writing the glue between the layers.  We can't avoid writing a byte cursor wrapper (see the comment about why we need non-owning bytebuf).

Adding iterators probably makes sense in a future step, and that may be the right interface to expose up to the SDK, but I'm reluctant to make that assumption until we have a better feel for usage patterns beyond the current set of callbacks and thin wrappers.

Reply:
I agree with Bret here. I'm not against iterators at all, and I think it would be a great thing for customers, but I want to see what the users actually do first. It may turn out they take the buffers and jam them into their own vectors or other structures immediately, then iterate.

Reply:
Granted. But I worry if you give them ByteCursor, they'll use it, even if there's a better approach lurking.
To get the same results you're mentioning would be to introduce iterators and see if it's not enough, then you would introduce a bytecursor. If you start with bytecursor, you will never have a need for iterators.